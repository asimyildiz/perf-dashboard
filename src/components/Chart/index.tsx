import { memo, FC } from 'react';
import { Line } from 'react-chartjs-2';
import moment from 'moment';

import { IChartData } from '../../interfaces/chartdata.interface';
import { IDevice } from '../../services/interfaces/device.interface';
import styles from './styles.module.scss';

/**
 * interface to represent the parameter
 * passed to the component
 * @interface Props
 */
interface Props {
  title: string;
  devices: IDevice[];
  items: IChartData[] | undefined;
}

/**
 * create data object which will be passed to the Line chart
 * @param {string} label - label of data (TTFB, FID...)
 * @param {Array<IDevice>} devices - device data
 * @param {Array<IChartData>|undefined} items - chart data array
 * @returns {unknown}
 */
const getChartData = (
  label: string,
  devices: IDevice[],
  items: IChartData[] | undefined,
): unknown => {
  if (items) {
    const data: number[] = items.map((item: IChartData) => item.value);
    const labels: string[][] = devices
      .filter(
        (device: IDevice) =>
          items.findIndex((item: IChartData) => item.id === device.id) > -1,
      )
      .map((device: IDevice) => {
        const date = moment(device.createdAt);
        return [date.format('MMMM Do YYYY'), date.format('h:mm:ss a')];
      });

    return {
      labels,
      datasets: [
        {
          label,
          data,
          fill: false,
          backgroundColor: styles.chartBackgroundColor,
          borderColor: styles.chartBorderColor,
        },
      ],
    };
  }

  return {};
};

const getOptions = () => ({
  responsive: true,
  plugins: {
    legend: {
      display: true,
      position: 'right',
      align: 'start',
    },
  },
  scales: {
    yAxes: [
      {
        ticks: {
          beginAtZero: true,
        },
      },
    ],
  },
});

/**
 * create a chart based on the data passed
 * @param {Props} param0 - props
 * @returns {JSXElement}
 */
const Chart: FC<Props> = ({ title, devices, items }: Props) => (
  <div className={styles.Chart}>
    <Line
      type="line"
      data={getChartData(title, devices, items)}
      options={getOptions()}
    />
  </div>
);

export default memo(Chart);
