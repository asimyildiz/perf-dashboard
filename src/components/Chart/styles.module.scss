$chart-background-color: rgb(255, 99, 132);
$chart-border-color: rgba(255, 99, 132, 0.2);

:export {
  chartBackgroundColor: $chart-background-color;
  chartBorderColor: $chart-border-color;
}

.Chart {
  margin: 5%;  
}
  
