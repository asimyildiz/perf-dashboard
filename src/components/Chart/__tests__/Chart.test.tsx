import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { fromVital } from "../../../interfaces/chartdata.interface";
import { IConnection } from "../../../services/interfaces/connection.interface";
import { IDevice } from "../../../services/interfaces/device.interface";
import { IVital } from "../../../services/interfaces/vital.interface";
import Chart from "../index";

const deviceData: IDevice[] = [
  {
    _id: "60d0431a74a8240af878195e",
    id: "v1-1624261119060-2428637885473",
    url: "http://localhost/kasaba/",
    referrer: "",
    userAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.106 Safari/537.36",
    memory: 8,
    cpus: 8,
    connection: {
      effectiveType: "4g",
      rtt: 50,
      downlink: 10,
    } as IConnection,
    createdAt: "2021-06-21T07:43:22.700+00:00"
  } as IDevice,
];

const vitalData: IVital[] = [
  {
    _id: "60d0431a74a8240af8781960",
    id: "v1-1624261119060-2428637885473",
    name: "FCP",
    delta: 1449.5999999940395,
    value: 1449.5999999940395,
  } as IVital,
  {
    _id: "60d0431a74a8240af8781961",
    id: "v1-1624261119060-2428637885473",
    name: "TTFB",
    delta: 353.7999999821186,
    value: 353.7999999821186,
  } as IVital,
  {
    _id: "60d0462a2fa99e0c816bcec7",
    id: "v1-1624262142614-3321620775826",
    name: "CLS",
    delta: 0.05607146357068885,
    value: 0.05607146357068885,
  } as IVital,
];

describe("<Chart />", () => {
  it("renders", () => {
    const tree = render(
      <MemoryRouter>
        <Chart
          title="TTFB"
          devices={deviceData}
          items={fromVital(vitalData)}
        />
      </MemoryRouter>
    ).container.firstChild;

    expect(tree).toMatchSnapshot();
  });
});
