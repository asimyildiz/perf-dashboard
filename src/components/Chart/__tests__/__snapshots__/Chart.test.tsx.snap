// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> renders 1`] = `
<div
  class="Chart"
/>
`;
